File: 2020-11-11/fusio.cc
#include <iostream>
#include <vector>
using namespace std;

// Retorna la fusió ordenada de dos vectors ordenats.
// precondició: v1 i v2 estàn ordenats.
vector<int> fusio(const vector<int>& v1, const vector<int>& v2)
{
    vector<int> v;
    int i = 0;
    int j = 0;
    int n1 = v1.size();
    int n2 = v2.size();
    while (i < n1 and j < n2) {
        if (v1[i] <= v2[j]) {
            v.push_back(v1[i]);
            ++i;
        } else {
            v.push_back(v2[j]);
            ++j;
        }
    }
    while (j < n2) {
        v.push_back(v2[j]);
        ++j;
    }
    while (i < n1) {
        v.push_back(v1[i]);
        ++i;
    }
    return v;
}

int main()
{
    vector<int> v1 = { 1, 5, 5, 7, 8, 9 };
    vector<int> v2 = { 2, 3, 5, 11 };
    vector<int> v = fusio(v1, v2);
    for (int x : v) {
        cout << x << endl;
    }
}

File: 2020-11-11/fusio.py

def fusio(v1, v2):
    """retorna la fusió ordenada de dos vectors ordenats"""
    n1 = len(v1)
    n2 = len(v2)
    i = 0
    j = 0
    v = []
    while i < n1 and j < n2:
        if v1[i] <= v2[j]:
            v.append(v1[i])
            i = i + 1
        else:
            v.append(v2[j])
            j = j + 1
    v.extend(v1[i:])
    v.extend(v2[j:])
    return v


def main():
    v1 = [1, 3, 4, 5, 6, 6]
    v2 = [2, 3, 5, 10, 11, 12]
    print(fusio(v1, v2))


main()
